python/pyspark/sql/connect/functions.py

```diff
@@ -80,6 +84,78 @@ def _invoke_binary_math_function(name: str, col1: Any, col2: Any) -> Column:
     return _invoke_function(name, *_cols)
 
 
+def _get_lambda_parameters(f: Callable) -> ValuesView[inspect.Parameter]:
+    signature = inspect.signature(f)
+    parameters = signature.parameters.values()
+
+    # We should exclude functions that use, variable args and keyword argument
+    # names, as well as keyword only args.
+    supported_parameter_types = {
+        inspect.Parameter.POSITIONAL_OR_KEYWORD,
+        inspect.Parameter.POSITIONAL_ONLY,
+    }
+
+    # Validate that the function arity is between 1 and 3.
+    if not (1 <= len(parameters) <= 3):
+        raise ValueError(
+            "f should take between 1 and 3 arguments, but provided function takes {}".format(
+                len(parameters)
+            )
+        )
+
+    # Verify that all arguments can be used as positional arguments.
+    if not all(p.kind in supported_parameter_types for p in parameters):
+        raise ValueError("All arguments of f must be usable as POSITIONAL arguments")
+
+    return parameters
+
+
+def _create_lambda(f: Callable) -> LambdaFunction:
+    """
+    Create `o.a.s.sql.expressions.LambdaFunction` corresponding
+    to transformation described by f
+
+    :param f: A Python of one of the following forms:
+            - (Column) -> Column: ...
+            - (Column, Column) -> Column: ...
+            - (Column, Column, Column) -> Column: ...
+    """
+    parameters = _get_lambda_parameters(f)
+
+    arg_names = ["x", "y", "z"][: len(parameters)]
+    arg_cols = [column(arg) for arg in arg_names]
+
+    result = f(*arg_cols)
```

Comment:
IIUC, we use `Column` with `UnresolvedAttribute` to invoke the function, then at the server side, we replace `UnresolvedAttribute` with `LambdaVariable`.

Shall we use `Column` with `LambdaVariable` to invoke the function in client side directly?

Reply:
the first commit used `UnresolvedNamedLambdaVariable` 
```
    arg_cols: List[Column] = []
    for arg in arg_names[: len(parameters)]:
        # TODO: How to make sure lambda variable names are unique? RPC for increasing ID?
        _uuid = str(uuid.uuid4()).replace("-", "_")
        arg_cols.append(Column(UnresolvedNamedLambdaVariable([f"{arg}_{_uuid}"])))

    result = f(*arg_cols)

    if not isinstance(result, Column):
        raise ValueError(f"Callable {f} should return Column, got {type(result)}")

    return LambdaFunction(result._expr, [arg._expr for arg in arg_cols])
```

then switched to `UnresolvedAttribute` according to the comments https://github.com/apache/spark/pull/39068#discussion_r1049561414

Reply:
I don't think this can simplify the client-side implementation, but it does save one proto message. At least we should document this in the proto message `LambdaFunction`: the function body should use `UnresolvedAttribute` as arguments to build the query plan.

Reply:
done